### ISSUE TITLE
Uncaught TypeError: patch.getChanges is not a function

### ISSUE THREAD

**Issue by OJFord** [atom/atom#11234]
1. Was working on (existing, gfm) document for some time
2. Error occurred on hitting enter (not for the first time)

Now every time causes this error. ~~Pasting (or vim putting) a `\n` character is fine, though; it's the actual keypress that is triggering a (attempted) call to `patch.getChanges`.~~ (That seemed to be true for a while; no longer the case.)

It's now also happening with other characters; mostly space.

Seemingly related (I imagine - without looking at the source - it needs the result of `getChanges`) the cursor is now positioned _after_ e.g. a closing paren, rather than inside the pair, on typing an open.

**Atom Version**: 1.7.0-beta0
**System**: Mac OS X 10.11.3
**Thrown From**: Atom Core

### Stack Trace

Uncaught TypeError: patch.getChanges is not a function

```
At /opt/homebrew-cask/Caskroom/atom-beta/1.6.0-beta3/Atom Beta.app/Contents/Resources/app.asar/node_modules/text-buffer/lib/text-buffer.js:1333

TypeError: patch.getChanges is not a function
    at TextBuffer.module.exports.TextBuffer.emitDidChangeTextEvent (/opt/homebrew-cask/Caskroom/atom-beta/1.6.0-beta3/Atom Beta.app/Contents/Resources/app.asar/node_modules/text-buffer/lib/text-buffer.js:1333:60)
    at TextBuffer.module.exports.TextBuffer.transact (/opt/homebrew-cask/Caskroom/atom-beta/1.6.0-beta3/Atom Beta.app/Contents/Resources/app.asar/node_modules/text-buffer/lib/text-buffer.js:843:12)
    at TextEditor.module.exports.TextEditor.transact (/opt/homebrew-cask/Caskroom/atom-beta/1.6.0-beta3/Atom Beta.app/Contents/Resources/app.asar/src/text-editor.js:1274:26)
    at atom-text-editor.newCommandListeners.(anonymous function) (/opt/homebrew-cask/Caskroom/atom-beta/1.6.0-beta3/Atom Beta.app/Contents/Resources/app.asar/src/register-default-commands.js:648:22)
    at CommandRegistry.module.exports.CommandRegistry.handleCommandEvent (/opt/homebrew-cask/Caskroom/atom-beta/1.6.0-beta3/Atom Beta.app/Contents/Resources/app.asar/src/command-registry.js:260:29)
    at /opt/homebrew-cask/Caskroom/atom-beta/1.6.0-beta3/Atom Beta.app/Contents/Resources/app.asar/src/command-registry.js:3:61
    at KeymapManager.module.exports.KeymapManager.dispatchCommandEvent (/opt/homebrew-cask/Caskroom/atom-beta/1.6.0-beta3/Atom Beta.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:570:16)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/opt/homebrew-cask/Caskroom/atom-beta/1.6.0-beta3/Atom Beta.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:378:22)
    at WindowEventHandler.module.exports.WindowEventHandler.handleDocumentKeyEvent (/opt/homebrew-cask/Caskroom/atom-beta/1.6.0-beta3/Atom Beta.app/Contents/Resources/app.asar/src/window-event-handler.js:98:36)
    at HTMLDocument.<anonymous> (/opt/homebrew-cask/Caskroom/atom-beta/1.6.0-beta3/Atom Beta.app/Contents/Resources/app.asar/src/window-event-handler.js:3:61)
```

### Commands

```
     -5:38.1.0 spell-check:correct-misspelling (atom-text-editor.editor.vim-mode.insert-mode.is-focused)
  3x -5:34.2.0 core:move-up (atom-text-editor.editor.vim-mode.insert-mode.is-focused)
     -5:08.9.0 autocomplete-plus:confirm (atom-text-editor.editor.vim-mode.insert-mode.is-focused.autocomplete-active)
     -5:08.6.0 editor:newline (atom-text-editor.editor.vim-mode.insert-mode.is-focused)
  2x -5:08.2.0 core:backspace (atom-text-editor.editor.vim-mode.insert-mode.is-focused)
     -5:07.8.0 editor:newline (atom-text-editor.editor.vim-mode.insert-mode.is-focused)
  3x -4:48.2.0 core:backspace (atom-text-editor.editor.vim-mode.insert-mode.is-focused.autocomplete-active)
  2x -3:38.3.0 editor:newline (atom-text-editor.editor.vim-mode.insert-mode.is-focused)
  2x -3:25.8.0 core:backspace (atom-text-editor.editor.vim-mode.insert-mode.is-focused)
  3x -3:19.1.0 core:move-right (atom-text-editor.editor.vim-mode.insert-mode.is-focused.autocomplete-active)
  2x -3:17.2.0 core:move-down (atom-text-editor.editor.vim-mode.insert-mode.is-focused)
     -2:09.3.0 pdf-view:zoom-out (div.pdf-view)
     -1:26.6.0 core:move-up (atom-text-editor.editor.vim-mode.insert-mode.is-focused)
  4x -0:52.9.0 core:backspace (atom-text-editor.editor.vim-mode.insert-mode.is-focused.autocomplete-active)
     -0:03.5.0 core:move-right (atom-text-editor.editor.vim-mode.insert-mode.is-focused.autocomplete-active)
  2x -0:02.3.0 editor:newline (atom-text-editor.editor.vim-mode.insert-mode.is-focused)
```

### Config

```json
{
  "core": {
    "closeEmptyWindows": false,
    "destroyEmptyPanes": false,
    "disabledPackages": [
      "markdown-scroll-sync"
    ],
    "excludeVcsIgnoredPath": true,
    "projectHome": "~",
    "themes": [
      "atom-dark-ui",
      "solarized-dark-syntax"
    ]
  },
  "editor": {
    "autoIndent": true,
    "fontFamily": "Anonymous Pro for Powerline",
    "fontSize": 11,
    "invisibles": {},
    "showIndentGuide": true,
    "showInvisibles": true,
    "showLineNumbers": true,
    "softTabs": true,
    "softWrap": true,
    "softWrapAtPreferredLineLength": true,
    "tabLength": 4,
    "tabType": "auto"
  }
}
```

### Installed Packages

```coffee
# User
markdown-preview-plus, v2.2.2
pdf-view, v0.50.0
tabs-to-spaces, v1.0.2
term3, v0.21.4
vim-mode, v0.65.0

# Dev
No dev packages
```

**Comment by joushou** [comment]
Same case here (new markdown document). Every newline suddenly caused that error. Closing the document and  opening it again made it go away, if at least temporarily.

**Comment by OJFord** [comment]
@joushou Are you also using `vim-mode`?

**Comment by joushou** [comment]
No vim-mode. If we ignore my syntax theme, go-plus and its dependencies, I'm working from a completely stock atom.

**Comment by ChrisJefferson** [comment]
I've got the same thing happening, no `vim-mode` either, editing a markdown file.

**Comment by jfchevrette** [comment]
Same error, no ```vim-mode```. I was editing a shell script.

**Comment by joushou** [comment]
It shows up after a while when typing markdown files rather consistently, and doesn't only show up on return. It is incredibly annoying, as the popups fill your screen.

After a short bit of debugging, the issue appears to be here:
(text-buffer.js:840)
```javascript
      compactedChanges = this.history.groupChangesSinceCheckpoint(checkpointBefore, endMarkerSnapshot, true);
      this.history.applyGroupingInterval(groupingInterval);
      this.emitMarkerChangeEvents(endMarkerSnapshot);
      this.emitDidChangeTextEvent(compactedChanges);
```

this.history.groupChangesSinceCheckpoint returns false, which this.emitDidChangeTextEvent tries to call "getChanges" on, which obviously fails.

The reason this.history.groupChangesSinceCheckpoint returns zero, is that this.undoStack is an array of 0 length, skipping the checkpointIndex search, and returning false at line history.js:131.

I have quickly hacked out the call to patch.getChanges in emitMarkerChangeEvents in the debugger when a "false" patch is provided, but I suspect that the real issue lies in the fact that my undo buffer is empty at this point, which I suspect was not the intention. After all, I made a change (a newline), and should therefore have something to undo. If no other progress is made, I might try to dive a little deeper, but I have no effin' clue how these parts of Atom work.

For now, I shall retreat to my static type systems.

**Comment by Superbest** [comment]
Is there a workaround for this? I am now getting it with every keystroke, and the undo queue is indeed empty (Ctrl+Z does nothing) as joushou says.

Disabling the Markdown preview doesn't seem to help much.

**Comment by joushou** [comment]
The workaround is to reopen the affected file, which works for a while.

No better patch is currently available. We need to figure out why the undo buffer dies.

**Comment by Superbest** [comment]
After leaving my first comment I ended up closing and reopening Atom, and the problem has indeed not come back.

**Comment by joushou** [comment]
It keeps coming back for me, however.

Has anyone tried a recent master? I don't have a suitable toolchain.

**Comment by shadyyx** [comment]
Running on 1.7.beta started from CLI.
Editing markdown file (now 180 lines, will be probably 200 when I am finished).
Installed packages:
* go-plus + deps
* git-blame
* git-diff-popup
* git plus
* linter-swagger
* minimap
* minimap-autohide
* minimap-hide
* minimap-(git|split)-diff

Also **no** `vim-mode`...

**Comment by nathansobo** [comment]
This will be fixed soon in 1.7.0-beta3. Keep in mind if you're having trouble on beta in the future, you can always fall back to the stable channel until your issue is resolved. Thanks for your help everyone.
